ipfs-5-to-6/migration/config_conv.go

```diff
@@ -0,0 +1,154 @@
+package mg5
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"strings"
+)
+
+func convertFile(orig string, new string, convFunc func(ds map[string]interface{}) error) error {
+	in, err := os.Open(orig)
+	if err != nil {
+		return err
+	}
+	out, err := os.Create(new)
+	if err != nil {
+		return err
+	}
+	return convert(in, out, convFunc)
+}
+
+func convert(in io.Reader, out io.Writer, convFunc func(ds map[string]interface{}) error) error {
+	data, err := ioutil.ReadAll(in)
+	if err != nil {
+		return err
+	}
+	conf := make(map[string]interface{})
+	err = json.Unmarshal(data, &conf)
+	if err != nil {
+		return err
+	}
+	fromLC := lcMap(conf)
+	ds := conf[fromLC["datastore"]].(map[string]interface{})
+	err = convFunc(ds)
+	if err != nil {
+		return err
+	}
+	conf[fromLC["datastore"]] = ds
+	fixed, err := json.MarshalIndent(conf, "", "  ")
+	if err != nil {
+		return err
+	}
+	out.Write(fixed)
+	out.Write([]byte("\n"))
+	return nil
+}
+
+func ver5to6(ds map[string]interface{}) error {
+	fromLC := lcMap(ds)
+	noSyncVal, _ := ds[fromLC["nosync"]]
+	if noSyncVal == nil {
+		noSyncVal = interface{}(false)
+	}
+	noSync, ok := noSyncVal.(bool)
+	if !ok {
+		return fmt.Errorf("unsupported value for Datastore.NoSync fields: %v", noSyncVal)
+	}
+	delete(ds, fromLC["nosync"])
+
+	dsTypeVal, _ := ds[fromLC["type"]]
+	if dsTypeVal == nil {
+		dsTypeVal = interface{}("")
+	}
+	dsType, ok := dsTypeVal.(string)
+	if !ok || (dsType != "default" && dsType != "leveldb" && dsType != "") {
+		return fmt.Errorf("unsupported value for Datastore.Type fields: %s", dsType)
+	}
+	delete(ds, fromLC["type"])
+
+	// Path and Params never appear to have been used so just delete them
+	delete(ds, fromLC["path"])
+	delete(ds, fromLC["params"])
+
+	ds["Spec"] = DatastoreSpec(!noSync)
+	return nil
+}
+
+// lcMap create a map of an all lowecase version as the keys in map as
+// the JSON library is case insensitive when matching json values to
+// struct fields
+func lcMap(conf map[string]interface{}) map[string]string {
+	fromLC := make(map[string]string)
+	for key, _ := range conf {
+		fromLC[strings.ToLower(key)] = key
+	}
+	return fromLC
+}
+
+func ver6to5(ds map[string]interface{}) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("incompatible config detected, downgrade not possible: %v", r.(error))
+		}
+	}()
+	fromLC := lcMap(ds)
+	spec := ds[fromLC["spec"]].(map[string]interface{})
```

Comment:
@whyrusleeping I am caching the panic to avoid having to write error checking code for every bit of the conversion. Do you just want a better error message if the "spec" field is missing or the wrong type?